src/main/java/edu/harvard/iq/dataverse/externaltools/ExternalToolHandler.java

```diff
@@ -40,52 +62,170 @@ public ExternalToolHandler(ExternalTool externalTool, DataFile dataFile, ApiToke
+                    paramsString = paramsString + (paramsString.isEmpty() ? "?" : "&") + entry.getKey() + "=";
+                    JsonValue val = entry.getValue();
+                    if (val.getValueType().equals(JsonValue.ValueType.NUMBER)) {
+                        paramsString += ((JsonNumber) val).intValue();
+                    } else {
+                        paramsString += ((JsonString) val).getString();
+                    }
                 }
-            });
-        });
-        if (!preview) {
-            return "?" + String.join("&", params);
+            } else {
+                //Send a signed callback to get params and signedURLs 
+                String callback = null;
+                switch (externalTool.getScope()) {
+                case DATASET:
+                    callback=SystemConfig.getDataverseSiteUrlStatic() + "/api/v1/datasets/"
+                            + dataset.getId() + "/versions/:latest/toolparams/" + externalTool.getId();
+                case FILE:
+                    callback= SystemConfig.getDataverseSiteUrlStatic() + "/api/v1/files/"
+                            + dataFile.getId() + "/metadata/" + fileMetadata.getId() + "/toolparams/"
+                            + externalTool.getId();
+                }
+                if (apiToken != null) {
+                    callback = UrlSignerUtil.signUrl(callback, 5, apiToken.getAuthenticatedUser().getUserIdentifier(), HttpMethod.GET,
+                            System.getProperty(SystemConfig.API_SIGNING_SECRET, "") + apiToken.getTokenString());
+                }
+                paramsString= "?callback=" + Base64.getEncoder().encodeToString(StringUtils.getBytesUtf8(callback));
+                if (getLocaleCode() != null) {
+                    paramsString += "&locale=" + getLocaleCode();
+                }
+            }
+            if (preview) {
+                paramsString += "&preview=true";
+            }
+            logger.fine("GET return is: " + paramsString);
+            return paramsString;
+
         } else {
-            return "?" + String.join("&", params) + "&preview=true";
+            // ToDo - if the allowedApiCalls() are defined, could/should we send them to
+            // tools using GET as well?
+
+            if (requestMethod.equals(HttpMethod.POST)) {
+                String body = JsonUtil.prettyPrint(createPostBody(params).build());
+                try {
+                    logger.info("POST Body: " + body);
+                    return postFormData(body);
+                } catch (IOException | InterruptedException ex) {
+                    Logger.getLogger(ExternalToolHandler.class.getName()).log(Level.SEVERE, null, ex);
+                }
+            }
         }
+        return null;
+    }
+
+    public JsonObject getParams(JsonObject toolParameters) {
+        //ToDo - why an array of object each with a single key/value pair instead of one object?
```

Comment:
Hmm. Good question. I agree. A single object like this probably makes more sense...

    "allowedApiCalls": [
      {
        "name":"retrieveDataFile",
        "httpMethod":"GET",
        "urlTemplate":"/api/v1/access/datafile/{fileId}",
        "timeOut":270
      }
    ]

As the TODO says, what we have now is this:

    "queryParameters": [
      {
        "fileid": "{fileId}"
      },
      {
        "datasetPid": "{datasetPid}"
      },
      {
        "locale":"{localeCode}"
      }
    ]

I guess I just didn't think of it.